### ISSUE TITLE
Fix indexing issues for pages accessible at multiple URLs

### ISSUE THREAD

**Issue by excentrickristy** [jakartaee/specifications#582]
Google Search Console is having trouble indexing specification pages because they are accessible with and without the .html extension with no canonical URL set. 

Example:

https://jakarta.ee/specifications/faces/3.0/jakarta-faces-3.0 
https://jakarta.ee/specifications/faces/3.0/jakarta-faces-3.0.html 

The jakarta.ee site is linking to the document using the .html extension, so it makes sense to use the version with the extension as the canonical version. But, since the rest of the site uses a trailing slash schema for the urls, it might actually be easier to rewrite the .html URLs to that schema. Then, run a search and replace for the links containing the .html urls in the specification pages to strengthen the internal linking structure.

Open to discussion on how to best deal with this one. At the very least we should be setting canonical URLs for the spec pages.

**Comment by excentrickristy** [comment]
@chrisguindon @ivargrimstad

Hi guys,
I was just reviewing these old issues and I know in the other issue, Ivar mentioned that I would have to make requests in each individual spec. Is that true for this as well? I would think this could be addressed on the server level?

Would love to get a better understanding.

**Comment by chrisguindon** [comment]
@excentrickristy I added to my queue an action item to build the site and investigate what is causing this duplication. 

I might not have time to get to this this week.

**Comment by chrisguindon** [comment]
Update: There is a setting in Netlify that is enabled by default called Pretty URLs:
https://docs.netlify.com/site-deploys/post-processing/

I've disabled it on our test site for the specification repo to see if this will fix the duplication issue. The testing website is under https://jakartaee-specifications.netlify.app/

**Comment by chrisguindon** [comment]
The change didn't seem to have fixed the issue. At least, from my end...

My next step is to download all the files that are deployed on production to confirm that the server does not have any duplicated files that could explain this duplication.

However, the download link on the Netlify site is falling. I will be contacting support about this issue.

**Comment by chrisguindon** [comment]
![Screenshot 2023-10-10 141958](https://github.com/jakartaee/specifications/assets/893942/9e18272f-cb5f-46be-a9bb-8d085f57e1a2)

I got access to the deployed files and there is no duplicate. This tells me that the server is creating these duplicates URL. I have pretty URLs disabled for both the spec repo and jakarta.ee.


I suspect there might be a bug where the pretty URLs setting is always on. I will ask them. If this does work, we might be required to migrate the website to EF preview framework where we would have more control over the server.

**Comment by chrisguindon** [comment]
I heard back from Netlify. Their response is that their CDN normalizes those URLs. .html and without .html and both served by them. This always worked this way and there are no plans to change this.

However, they did mention that we could look at using Edge Functions: [Edge Functions overview | Netlify Docs](https://docs.netlify.com/edge-functions/overview/) to add the canonical header or to even simply redirect to a particular version.

**Comment by chrisguindon** [comment]
Disabling pretty URLs on jakarta.ee triggered a regression on our spec pages:
https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/3838

**Comment by excentrickristy** [comment]
@chrisguindon - dredging up an old issue here. Since pretty URLS do not seem to work, can we set a canonical tag on these pages?

**Comment by excentrickristy** [comment]
That sounds good. Since the main website links to the version of the page without .html, that's probably the URL we should use a the canonical URL.

Additionally, it would be nice to do a templated title tag to include the version number as well as the word specification.

**Comment by chrisguindon** [comment]
> Additionally, it would be nice to do a templated title tag to include the version number as well as the word specification.

@ivargrimstad can we ask our spec projects to follow a template for how set their title tags for their docs? What would be the best way for @excentrickristy to ask them that.

**Comment by ivargrimstad** [comment]
That should be possible if the asciidoctor plugin supports it. The documents are all generated from AsciiDoc. Would only be for future documents.

**Comment by chrisguindon** [comment]
Thanks @ivargrimstad 

@excentrickristy I don't think I need to be involved in that conversation but I would suggest that you start with a new issue asking spec projects to follow a title convention for their future documents!

Regarding the content duplication issue, I just deployed this change:
https://github.com/jakartaee/jakarta.ee/commit/db32dce46dbeb8fdfcfd769bf4c1220d5860656a

This add a new HTTP header for both  https://jakarta.ee/specifications/faces/3.0/jakarta-faces-3.0.html and  https://jakarta.ee/specifications/faces/3.0/jakarta-faces-3.0 that says:

```
❯ curl -I https://jakarta.ee/specifications/faces/3.0/jakarta-faces-3.0.html
HTTP/2 200
accept-ranges: bytes
age: 1
cache-control: public,max-age=0,must-revalidate
cache-status: "Netlify Edge"; fwd=miss
...
link: <https://www.jakarta.ee/specifications/faces/3.0/jakarta-faces-3.0>; rel="canonical"
...
```


Given the current constraints, I believe this is the best we can do.

Should you require additional configurations, please submit a request by including the details in the same format as my commit within an issue against the Jakarta EE website. 

For further details, refer to the official documentation on custom headers from Netlify:
https://docs.netlify.com/routing/headers/

Let me know if you have any questions, otherwise I am thinking we can now close this!

**Comment by excentrickristy** [comment]
@chrisguindon hmmm I'm not seeing those changes on live. the faces spec pages don't appear to have canonicals set?

**Comment by excentrickristy** [comment]
@chrisguindon ah ok I see the canonical in the headers but the issue is that it's using www.jakarta.ee as a root domain instead of jakarta.ee in the canonical link so both faces urls are coming up as non-indexable.

**Comment by chrisguindon** [comment]
@excentrickristy Good catch @excentrickristy - I made the fix with https://github.com/jakartaee/jakarta.ee/commit/6fd6676587143175404b4e82d87f35843b7180fc

I would expect to go live in the next 15 minutes.

**Comment by excentrickristy** [comment]
@chrisguindon Thank you - it's coming up properly now!

So - If I understand correctly, I will have to file an issue for each spec doc with the code from your commit but updated to reflect the proper URLS.

I will tag you in my first one to make sure I am doing it right! Thank you for your help.

**Comment by chrisguindon** [comment]
@excentrickristy - We can do a bunch of them at the same time. It would definitely simplify things if you can provide the changes in the same format as this file so that we can simply copy and paste what you need: https://github.com/jakartaee/jakarta.ee/blob/src/static/_headers

**Comment by chrisguindon** [comment]
However, we can start with updating one to make sure you have the correct format.
